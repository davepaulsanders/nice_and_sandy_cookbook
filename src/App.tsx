import { useState } from 'react'
import './App.css'

function App() {

  return (
    <div>
	<p className='text-4xl'>Hey</p>
    </div>
  )
}

export default App
