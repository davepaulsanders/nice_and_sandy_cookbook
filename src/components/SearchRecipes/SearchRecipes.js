"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var SearchRecipes = function (_a) {
    var setSearch = _a.setSearch, cachedRecipes = _a.cachedRecipes, recipes = _a.recipes, setRecipes = _a.setRecipes;
    var _b = (0, react_1.useState)(""), searchInputVal = _b[0], setSearchInputVal = _b[1];
    var _c = (0, react_1.useState)([]), filteredRecipes = _c[0], setFilteredRecipes = _c[1];
    var _d = (0, react_1.useState)(true), hideSearchList = _d[0], setHideSearchList = _d[1];
    var resetRecipes = function (e) {
        setSearch(false);
        setRecipes(cachedRecipes.current);
        setSearchInputVal("");
    };
    var searchRecipes = function (val) {
        var reg = new RegExp("".concat(val), "i");
        setRecipes(cachedRecipes.current.filter(function (recp) {
            if (reg.test(recp.label)) {
                return recp;
            }
            else {
                return;
            }
        }));
    };
    var handleChange = function (e) {
        setSearch(true);
        setSearchInputVal(e.target.value);
        searchRecipes(e.target.value);
        if (e.target.value === "") {
            setHideSearchList(true);
            setSearch(false);
        }
        else {
            setHideSearchList(false);
        }
    };
    var handleIndividualRecipeClick = function (recp) {
        setHideSearchList(true);
        setSearch(true);
        setRecipes([recp]);
        //setFilteredRecipes(cachedRecipes.current)
    };
    return (<div className="flex flex-col items-center">
			<input placeholder="Search recipes" className="relative rounded-md h-[3rem] w-[20rem] py-[8px] mt-4 text-xl pl-2" type="text" value={searchInputVal} onChange={handleChange}/>
			{!hideSearchList && (<ul className="absolute z-[1000] top-[285px]">
								 {recipes.map(function (recp) { return <li className="py-2 hover:bg-slate-200 hover:cursor-pointer bg-white w-[20rem] mx-auto" onClick={function () { return handleIndividualRecipeClick(recp); }}>{recp.label}</li>; })}
			</ul>)}
			<button className="mt-6 bg-tan-500 text-black rounded-md border py-2 px-8 bg-medg border-1 border-lightg" onClick={resetRecipes}>Reset</button>
		</div>);
};
exports.default = SearchRecipes;
