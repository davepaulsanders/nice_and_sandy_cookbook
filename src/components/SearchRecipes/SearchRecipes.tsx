import { Recipe } from "../../types/types";
import { useState, useEffect } from "react";
interface SearchRecipesProps {
	recipes: Recipe[];
	setSearch: React.Dispatch<React.SetStateAction<boolean>>;
	setRecipeSearchCopy: React.Dispatch<React.SetStateAction<Recipe[]>>;
}
const SearchRecipes = ({
	recipes,
	setSearch,
	setRecipeSearchCopy,
}: SearchRecipesProps) => {
	const [searchInputVal, setSearchInputVal] = useState("");

	// if recipes changes, refilter with changes
	useEffect(() => {
		const reg = new RegExp(`${searchInputVal}`, "i");
		const filtered = recipes.filter((recp) => reg.test(recp.label));
		// don't trigger if searchInputVal is empty
		if (searchInputVal === "") {
			return;
		}
		setRecipeSearchCopy(filtered);
	}, [recipes, searchInputVal]);

	const resetRecipes = async (e) => {
		setSearch(false);
		setSearchInputVal("");
		setRecipeSearchCopy(recipes);
	};
	const handleInputChange = (e) => {
		// clear if input is empty
		if (!e.target.value) {
			resetRecipes(e);
		} else {
			setSearch(true);
			setSearchInputVal(e.target.value);
		}
	};
	return (
		<div className="flex flex-col items-center">
			<input
				placeholder="Search recipes"
				className="relative rounded-md h-[3rem] w-[20rem] py-[8px] mt-4 text-xl pl-2"
				type="text"
				value={searchInputVal}
				onChange={handleInputChange}
			/>
			<button
				className="mt-6 bg-tan-500 text-black rounded-md border py-2 px-8 bg-medg border-1 border-lightg"
				onClick={resetRecipes}
			>
				Reset
			</button>
		</div>
	);
};
export default SearchRecipes;
