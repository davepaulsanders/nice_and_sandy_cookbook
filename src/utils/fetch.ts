export const fetchData = async (url: string) => {
	const resp = await fetch(url)
	if (resp.status !== 200) throw new Error(`Request failed with ${resp.status}`)
	const data = resp.json()
	return data
}
