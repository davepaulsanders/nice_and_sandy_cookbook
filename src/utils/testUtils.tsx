import RecipeCard from '../components/RecipeCard/RecipeCard';
import { ReactElement } from "react"
import { vi } from "vitest"
import { Recipe } from "../types/types"

const generateRecipes = () => {
const recipes: Recipe[] = []
for (let i = 0; i < 10; i++) {
	recipes.push({
	id:i + 1,
	img:"https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg",
	href:"https://www.feastingathome.com/vegan-fajitas/",
	label:`Vegan Fajitas ${i + 1}`,
	alt:"test",
	category:"ENTREE",
	category_id:1,
	is_pinned: false,
	})
}
return recipes
}
const generateRecipeCards = () => {
const recipeCards: ReactElement[] = []
for (let i = 0; i < 10; i++) {
	recipeCards.push(
	<RecipeCard
	id={i + 1}
	img="https://www.feastingathome.com/wp-content/uploads/2024/10/Vegan-Fajitas-13.jpg"
	href="https://www.feastingathome.com/vegan-fajitas/"
	label="Vegan Fajitas 2"
	alt="test"
	category="ENTREE"
	category_id={1}
	is_pinned= {false}
	recipes={generateRecipes()}
	setRecipes={vi.fn()}
	/>)
	}
	return recipeCards
}
export { generateRecipeCards, generateRecipes }
