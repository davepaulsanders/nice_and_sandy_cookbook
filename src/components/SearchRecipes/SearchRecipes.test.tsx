import SearchRecipes from "./SearchRecipes";
import App from "../../App";
import { expect, test, vi } from "vitest";
import { render, waitFor, screen } from "@testing-library/react";

import { generateRecipes } from "../../utils/testUtils";

const recipes = generateRecipes();

import "@testing-library/dom";
import userEvent from "@testing-library/user-event";

global.fetch = vi.fn((url: string) => {
  if (url === '../v1/recipes') {
    return Promise.resolve({
	status: 200,
      json: () => Promise.resolve({ recipes: generateRecipes() }),
    });
  }
  if (url === '../v1/categories') {
    return Promise.resolve({
	status: 200,
      json: () => Promise.resolve({ categories: ["ENTREE"] }),
    });
  }
  return Promise.resolve({
    json: () => Promise.resolve({ error: 'Unknown path' }),
  });
}) as unknown as typeof global.fetch

test("<SearchRecipes /> renders", () => {
	const { container } = render(
		<SearchRecipes
			recipes={recipes}
			setSearch={vi.fn()}
			setRecipeSearchCopy={vi.fn()}
		/>
	);
	expect(container).toMatchSnapshot();
});
test("<Calls setSearch with false if input is cleared, and true if input has text", async () => {
	const setSearchMock = vi.fn()
	render(
		<SearchRecipes
			recipes={recipes}
			setSearch={setSearchMock}
			setRecipeSearchCopy={vi.fn()}
		/>
	);
	const searchInput = await waitFor(() => screen.getByPlaceholderText("Search recipes"));
	await waitFor(() => userEvent.click(searchInput));
	await waitFor(() => userEvent.keyboard("a"));
	expect(setSearchMock).toBeCalledWith(true)
	expect(setSearchMock).toBeCalledTimes(1)
	await waitFor(() => userEvent.keyboard("{backspace}"));
	expect(setSearchMock).toBeCalledWith(false)
	expect(setSearchMock).toBeCalledTimes(2)
});
test("Reset button clears input", async () => {
	const setSearchMock = vi.fn()
	render(
		<SearchRecipes
			recipes={recipes}
			setSearch={setSearchMock}
			setRecipeSearchCopy={vi.fn()}
		/>
	);
	const searchInput = await waitFor(() => screen.getByPlaceholderText("Search recipes"));
	await waitFor(() => userEvent.click(searchInput));
	await waitFor(() => userEvent.keyboard("a"));
	const resetButton = await waitFor(() => screen.getByText("Reset"));
	await waitFor(() => userEvent.click(resetButton));
	const searchInputAfterReset = await waitFor(() => screen.getByPlaceholderText("Search recipes"));
	expect(searchInputAfterReset).not.toBeNull()
});
test("Typing in input calls setRecipeSearchCopy", async () => {
	const setRecipeSearchCopyMock = vi.fn()
	render(
		<SearchRecipes
			recipes={recipes}
			setSearch={vi.fn()}
			setRecipeSearchCopy={setRecipeSearchCopyMock}
		/>
	);
	const searchInput = await waitFor(() => screen.getByPlaceholderText("Search recipes"));
	await waitFor(() => userEvent.click(searchInput));
	await waitFor(() => userEvent.keyboard("a"));
	expect(setRecipeSearchCopyMock).toBeCalledTimes(1)	
});
test("Recipes filter based on input", async () => {
	render( <App />);
	const searchInput = await waitFor(() => screen.getByPlaceholderText("Search recipes"));
	await waitFor(() => userEvent.click(searchInput));
	await waitFor(() => userEvent.keyboard("2"));
	const veganRecp = await waitFor(() => screen.getByText("Vegan Fajitas 2"));
	const veganRecp1 = await waitFor(() => screen.queryByText("Vegan Fajitas"));
	expect(veganRecp).not.toBeNull()
	expect(veganRecp1).toBeNull()
});
