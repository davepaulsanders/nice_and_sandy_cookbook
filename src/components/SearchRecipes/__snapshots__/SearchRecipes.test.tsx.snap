// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchRecipes /> renders 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <input
      class="relative rounded-md h-[3rem] w-[20rem] py-[8px] mt-4 text-xl pl-2"
      placeholder="Search recipes"
      type="text"
      value=""
    />
    <button
      class="mt-6 bg-tan-500 text-black rounded-md border py-2 px-8 bg-medg border-1 border-lightg"
    >
      Reset
    </button>
  </div>
</div>
`;
